### ISSUE TITLE
Unexpected nullability warning for pattern matching control flow

### ISSUE THREAD

**Issue by tr4cks** [dotnet/roslyn#65306]
**Version Used**: .net 6.0.402 c#10 on macOS m1

**Steps to Reproduce**:

I have reproduced a very small code example below. I have warnings telling me that the value can be null but I think it is not possible according to the conditions.
You can see [here](https://stackoverflow.com/questions/74369516/pattern-matching-condition-ensures-neither-property-is-null-but-compiler-issues) for more explaination.

```csharp
Console.WriteLine("Code needed to build the project");

public interface IAbove
{
    string? AboveValue { get; }
}

public interface IBelow : IAbove
{
    string? BelowValue { get; }
}

public class Test
{
    public static void Execute(IAbove above)
    {
        if (above is IBelow { AboveValue: not null, BelowValue: not null } belowTest1)
        {
            // warning CS8602: Dereferencing of a possible null reference
            Console.WriteLine(belowTest1.AboveValue.Length);
            // ok
            Console.WriteLine(belowTest1.BelowValue.Length);
        }
        if (above is not IBelow { AboveValue: not null, BelowValue: not null } belowTest2)
        {
            return;
        }
        // warning CS8602: Dereferencing of a possible null reference
        Console.WriteLine(belowTest2.AboveValue.Length);
        // ok
        Console.WriteLine(belowTest2.BelowValue.Length);
    }
}
```

**Expected Behavior**:

I think the code is valid, and there should be no nullability warning.

**Actual Behavior**:

There's a nullability warning.

**Comment by RikkiGibson** [comment]
Spent a little time investigating this bug.

Consider a sample like the following:
```cs
class Base
{
    public string? Prop { get; set; }
}

class Derived : Base { }

class Program
{
    void M(Base @base)
    {
        if (@base is Derived { Prop: not null } derived)
        {
            derived.Prop.ToString();
            @base.Prop.ToString();
        }
    }
}
```

The expression `@base is Derived { Prop: not null } derived)` results in the decision DAG, where `t0` is `@base`:

```
[0]: t0 is Derived ? [1] : [4]
[1]: t1 = (Derived)t0; [2]
[2]: t2 = t0.Prop; [3]
[3]: t2 == null ? [4] : [5]
[4]: leaf <isPatternFailure> `Derived { Prop: not null } derived`
[5]: when <true> ? [6] : <unreachable>
[6]: leaf <isPatternSuccess> `Derived { Prop: not null } derived`
```

Importantly, we read `t0.Prop` instead of reading `t1.Prop`. This basically means we use the slot for `@base.Prop` for the null check instead of `derived.Prop`. My gut feeling is we will want to recognize this as a special condition of some kind and go out of our way to copy the "test results" into the state for both `@base.Prop` and `derived.Prop`.

You could imagine this maybe happening at any level of recursion as well. `if (wrapper is { @base: Derived { Prop: not null } derived }`, etc. I don't *think* we can use a solution like the one we do for `a = b` assignments, where all the states within `b` are copied in to the state for `a`, because it would mean that states for properties only found in `a` would be lost.

We could perhaps make it so the tempMap can relate a DAG temp to more than one input variable, i.e. use `(int slot, TypeSymbol type)[]` instead of `(int slot, TypeSymbol type)` below. But it might complicate things a lot.

https://github.com/dotnet/roslyn/blob/e8f1b4026df958abb7c5dbe944679da9c64bc7b4/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs#L359
